### ISSUE TITLE
Unsound examples in the "Pinning" chapter

### ISSUE THREAD

**Issue by mexus** [rust-lang/async-book#117]
The `Test` examples (https://github.com/rust-lang/async-book/blame/641dc148b34a36d68a9d64047d4712d1a6e932ac/src/04_pinning/01_chapter.md#L136, https://github.com/rust-lang/async-book/blame/641dc148b34a36d68a9d64047d4712d1a6e932ac/src/04_pinning/01_chapter.md#L363)  and all its copy-pastes in the Pinning chapter provide unsound API: the potential user might easily call `Test::b` on a `Test` object that has not been previously initialized with a `Test::init` call.

If my understanding is correct, either `Test::b` should be marked as an `unsafe` method, or we should rethink the whole API (by introducing a safe macro-wrapper, for instance).

P.S.

Sorry for the links with "blame", but it seems like there is no way to make a link to an non-rendered markdown in github: https://github.com/isaacs/github/issues/297

**Comment by mexus** [comment]
Here's an example solution with the macro I've mentioned above: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7bdb3da6fa9b35e6e6b443daf92e1d29

TL;DR: we're declaring the `new` method as `unsafe` since it creates a `Test` instance, which is not safe to use, and provide a safe macro-wrapper that performs the proper initialization.

**Comment by taiki-e** [comment]
Good catch! I think the addition of `assert!(!self.b.is_null())` is a reasonable fix.
